bin/check-dns-zone.rb

```diff
@@ -0,0 +1,204 @@
+#
+# OUTPUT:
+#   plain text
+#
+# PLATFORMS:
+#   Linux, BSD
+#
+# DEPENDENCIES:
+#   gem: sensu-plugin
+#   gem: dnsruby
+#   gem: whois
+#
+# USAGE:
+#   example commands
+#
+# NOTES:
+#   Does it behave differently on specific platforms, specific use cases, etc
+#
+# LICENSE:
+#   Zubov Yuri <yury.zubau@gmail.com> sponsored by Actility, https://www.actility.com
+#   Released under the same terms as Sensu (the MIT license); see LICENSE
+#   for details.
+#
+
+require 'sensu-plugin/check/cli'
+require 'dnsruby'
+require 'whois'
+require 'whois/parser'
+require 'ipaddr'
+require 'resolv'
+
+#
+# DNS
+#
+class DNSZone < Sensu::Plugin::Check::CLI
+  option :domain,
+         description: 'Domain to resolve (or ip if type PTR)',
+         short: '-d DOMAIN',
+         long: '--domain DOMAIN'
+
+  option :threshold,
+         description: 'Percentage of DNS queries that must succeed',
+         short: '-l PERCENT',
+         long: '--threshold PERCENT',
+         proc: proc(&:to_i),
+         default: 100
+
+  option :timeout,
+         description: 'Set timeout for query',
+         short: '-T TIMEOUT',
+         long: '--timeout TIMEOUT',
+         proc: proc(&:to_i),
+         default: 5
+
+  option :warn_only,
+         description: 'Warn instead of critical on failure',
+         short: '-w',
+         long: '--warn-only',
+         boolean: true
+
+  def resolve_ns
+    Resolv::DNS.new.getresources(config[:domain], Resolv::DNS::Resource::IN::NS).map { |e| e.name.to_s }
+  end
+
+  def check_nss(entries)
+    errors = []
+    success = []
+    record = Whois.whois(config[:domain])
+    parser = record.parser
+    additional_text = "(whois #{parser.nameservers.map(&:name)}) (dig #{entries})"
+    if Set.new(parser.nameservers.map(&:name)) == Set.new(entries)
+      success << "Resolved #{config[:domain]} #{config[:type]} equal with whois #{additional_text}"
+    else
+      errors << "Resolved #{config[:domain]} #{config[:type]} did not include #{config[:result]} #{additional_text}"
+    end
+    [errors, success]
+  end
+
+  def check_results(entries)
+    errors = []
+    success = []
+
+    %w[check_nss check_udp check_tcp check_axfr soa_query].each do |check|
+      result = send(check, entries)
+      errors.push(*result[0])
+      success.push(*result[1])
+    end
+
+    [errors, success]
+  end
+
+  def check_udp(entries)
+    errors = []
+    success = []
+    entries.each do |ns|
+      Dnsruby::Resolv.getaddresses(ns).each do |ip|
+        begin
+          resolv = Dnsruby::Resolver.new(nameserver: ip.to_s, do_caching: false, use_tcp: false)
+          resolv.query_timeout = config[:timeout]
+          resolv.query(config[:domain], Dnsruby::Types.ANY)
+
+          success << "Resolved DNS #{ns}(#{ip}) uses udp"
+        rescue StandardError
+          errors << "Resolved DNS #{ns}(#{ip}) doesn't use udp"
+        end
+      end
+    end
+    [errors, success]
+  end
+
+  def check_tcp(entries)
```

Comment:
can we use a better name? how about something more like `check_dns_tcp`?

Reply:
Also I feel like we can combine the two checks and just pass in whether we want udp or tcp.

Reply:
@majormoses, I agree with you - these methods have the same checking logic.
I will merge them.